subcommand.go

```diff
@@ -0,0 +1,51 @@
+package main
+
+import (
+	"log"
+	"os"
+	"os/exec"
+)
+
+type SubCommand struct {
+	cmd      *exec.Cmd
+	PipeRead *os.File
+}
+
+type Stream int
+
+const (
+	StreamStdout Stream = 1 << iota
+	StreamStderr
+)
+
+func NewSubCommand(args []string, stream Stream) SubCommand {
+	pipeRead, pipeWrite, err := os.Pipe()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	env := append(os.Environ(), "GODEBUG=gctrace=1")
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Env = env
+	cmd.Stdin = os.Stdin
+
+	switch stream {
+	case StreamStdout:
+		cmd.Stdout = pipeWrite
```

Comment:
Thanks. I may not have been clear. What I see is when the command under test runs, any output that gets sent to stderr is lost.,the command I was testing was godoc.



> On 12 Oct 2015, at 19:38, Vincent Bonmalais <notifications@github.com> wrote:
> 
> In subcommand.go:
> 
> > +)
> > +
> > +func NewSubCommand(args []string, stream Stream) SubCommand {
> > +	pipeRead, pipeWrite, err := os.Pipe()
> > +	if err != nil {
> > +		log.Fatal(err)
> > +	}
> > +
> > +	env := append(os.Environ(), "GODEBUG=gctrace=1")
> > +	cmd := exec.Command(args[0], args[1:]...)
> > +	cmd.Env = env
> > +	cmd.Stdin = os.Stdin
> > +
> > +	switch stream {
> > +	case StreamStdout:
> > +		cmd.Stdout = pipeWrite
> I'm not sure I'm understanding this comment. The only difference between master and this branch is the switch statement (and I only added it to make it testable). In main.go, I'm still using StreamStderr, which means this should behave the same as master.
> 
> Speaking of failing commands, I can see now a race condition where s.cmd.Run() fails and log.Fatal(err) exits before the parser has time to finish its job. But that would apply to the master branch too, wouldn't it?
> 
> I will have another run to add tests around this scenario.
> 
> —
> Reply to this email directly or view it on GitHub.
>